apps/bfd-data-npi/src/main/java/gov/cms/bfd/data/npi/utility/DataUtilityCommons.java

```diff
@@ -0,0 +1,329 @@
+package gov.cms.bfd.data.npi.utility;
+
+import com.google.common.base.Strings;
+import java.io.BufferedOutputStream;
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.FilenameFilter;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.net.URL;
+import java.nio.charset.Charset;
+import java.nio.charset.CharsetDecoder;
+import java.nio.charset.CharsetEncoder;
+import java.nio.charset.CodingErrorAction;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Calendar;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+import org.apache.commons.io.FileUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Data Utility Commons class for npi. * */
+public class DataUtilityCommons {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DataUtilityCommons.class);
+
+  /** Size of the buffer to read/write data. */
+  private static final int BUFFER_SIZE = 4096;
+
+  /** The day of the month we should check to see if the file has posted. */
+  private static final int DAYS_IN_EXPIRATION = 10;
+
+  /**
+   * Gets the org names from the npi file.
+   *
+   * @param outputDir the output directory
+   * @param downloadUrl the downloadUrl passed in
+   * @param npiFile the npi file
+   */
+  public static void getNPIOrgNames(String outputDir, Optional<String> downloadUrl, String npiFile)
+      throws IOException, IllegalStateException {
+    Path outputPath = Paths.get(outputDir);
+    if (!Files.isDirectory(outputPath)) {
+      throw new IllegalStateException("OUTPUT_DIR does not exist for NPI download.");
+    }
+
+    // Create a temp directory that will be recursively deleted when we're done.
+    Path tempDir = null;
+    try {
+      tempDir = Files.createTempDirectory("npi-data");
+    } catch (IOException e) {
+      throw new IllegalStateException("Cannot create temporary directory.");
+    }
+
+    // If the output file isn't already there, go build it.
+    Path convertedNpiDataFile = outputPath.resolve(npiFile);
+    try {
+      buildNPIResource(convertedNpiDataFile, downloadUrl, tempDir);
+    } catch (IOException exception) {
+      LOGGER.error("NPI data file could not be read.  Error:", exception);
```

Comment:
same question on flow control -- we log an error but don't re-throw. How is the caller supposed to handle this?

Reply:
rethrowing error